overlord/clusterstate/clusterstate.go

```diff
@@ -23,26 +23,194 @@ import (
+	}
+
+	st.Set("cluster", clusterState{
+		Current: clusterAssertionState{
+			ClusterID:   cluster.ClusterID(),
+			Sequence:    cluster.Sequence(),
+			AuthorityID: cluster.AuthorityID(),
+		},
+	})
+
+	return nil
+}
+
+// UpdateCluster installs an incremental cluster assertion update. Callers must
+// hold the state lock.
+func UpdateCluster(st *state.State, bundle io.Reader) error {
+	batch, cluster, err := decodeClusterBundle(bundle)
+	if err != nil {
+		return err
+	}
+
+	var cs clusterState
+	if err := st.Get("cluster", &cs); err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return ErrNoClusterAssertion
+		}
+		return err
+	}
+
+	if cluster.AuthorityID() != cs.Current.AuthorityID {
+		return fmt.Errorf(
+			"cluster assertion authority %q does not match expected authority %q",
+			cluster.AuthorityID(),
+			cs.Current.AuthorityID,
+		)
+	}
+
+	if cluster.ClusterID() != cs.Current.ClusterID {
+		return fmt.Errorf(
+			"cluster assertion id %q does not match expected id %q",
+			cluster.ClusterID(),
+			cs.Current.ClusterID,
+		)
+	}
+
+	if cluster.Sequence() <= cs.Current.Sequence {
+		return fmt.Errorf(
+			"cluster assertion sequence %d must be greater than current sequence %d",
+			cluster.Sequence(),
+			cs.Current.Sequence,
+		)
+	}
+
+	if err := assertstate.AddBatch(st, batch, nil); err != nil {
+		return fmt.Errorf("cannot add cluster assertion bundle: %w", err)
+	}
+
+	st.Set("cluster", clusterState{
+		Current: clusterAssertionState{
+			ClusterID:   cluster.ClusterID(),
+			Sequence:    cluster.Sequence(),
+			AuthorityID: cluster.AuthorityID(),
+		},
+	})
+
+	return nil
+}
+
+// CurrentCluster returns the currently tracked cluster assertion. Callers must
+// hold the state lock.
+func CurrentCluster(st *state.State) (*asserts.Cluster, error) {
+	var cs clusterState
+	if err := st.Get("cluster", &cs); err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return nil, ErrNoClusterAssertion
+		}
+		return nil, err
+	}
+
+	headers := map[string]string{
+		"cluster-id": cs.Current.ClusterID,
+		"sequence":   strconv.Itoa(cs.Current.Sequence),
+	}
+	a, err := assertstate.DB(st).Find(asserts.ClusterType, headers)
+	if err != nil {
+		return nil, fmt.Errorf("cannot resolve cluster assertion: %w", err)
+	}
+
+	cluster, ok := a.(*asserts.Cluster)
+	if !ok {
+		return nil, errors.New("internal error: stored assertion is not a cluster assertion")
+	}
+	return cluster, nil
+}
+
+func decodeClusterBundle(bundle io.Reader) (*asserts.Batch, *asserts.Cluster, error) {
+	var cluster *asserts.Cluster
+	batch := asserts.NewBatch(nil)
+
+	dec := asserts.NewDecoder(bundle)
+	for {
+		a, err := dec.Decode()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot decode cluster assertion bundle: %w", err)
+		}
+
+		if err := batch.Add(a); err != nil {
+			return nil, nil, err
+		}
+
+		c, ok := a.(*asserts.Cluster)
+		if !ok {
+			continue
+		}
+		cluster = c
```

Comment:
it's probably an error if there was more than one cluster assertion in the bundle